go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -2033,6 +2035,228 @@ func (s *VtctldServer) UpdateCellsAlias(ctx context.Context, req *vtctldatapb.Up
+	si, err := s.ts.GetShard(ctx, req.Keyspace, req.Shard)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetShard(%v, %v) failed: %v", req.Keyspace, req.Shard, err)
+	}
+
+	// get schema from the master, or error
+	if !si.HasPrimary() {
+		return nil, vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no primary in shard %v/%v", req.Keyspace, req.Shard)
+	}
+	log.Infof("Gathering schema for primary %v", topoproto.TabletAliasString(si.Shard.PrimaryAlias))
+
+	getTabletReq := &vtctldata.GetTabletRequest{
+		TabletAlias: si.Shard.PrimaryAlias,
+	}
+
+	getTabletResponse, err := s.GetTablet(ctx, getTabletReq)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetTablet(%v) failed: %v", getTabletReq.TabletAlias, err)
+	}
+
+	masterSchema, err := s.tmc.GetSchema(ctx, getTabletResponse.Tablet, nil, req.ExcludeTables, req.IncludeViews)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetSchema(%v, nil, %v, %v) failed: %v", si.Shard.PrimaryAlias, req.ExcludeTables, req.IncludeViews, err)
+	}
+
+	if req.IncludeVSchema {
+		validateVSchemaReq := &vtctldatapb.ValidateVSchemaRequest{
+			Keyspace:      req.Keyspace,
+			Shards:        []string{req.Shard},
+			ExcludeTables: req.ExcludeTables,
+			IncludeViews:  req.IncludeViews,
+		}
+
+		if _, err := s.ValidateVSchema(ctx, validateVSchemaReq); err != nil {
+			return nil, vterrors.Wrapf(err, "ValidateVSchema(%v) failed: %v", validateVSchemaReq, err)
+		}
+	}
+
+	// read all the aliases in the shard, that is all tablets that are
+	// replicating from the master
+	aliases, err := s.ts.FindAllTabletAliasesInShard(ctx, req.Keyspace, req.Shard)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "FindAllTabletAliasesInShard(%v, %v) failed: %v", req.Keyspace, req.Shard, err)
+	}
+
+	// then diff with all replicas
+	var (
+		rec concurrency.AllErrorRecorder
+		wg  sync.WaitGroup
+	)
+
+	for _, alias := range aliases {
+		if topoproto.TabletAliasEqual(alias, si.Shard.PrimaryAlias) {
+			continue
+		}
+
+		wg.Add(1)
+		go s.diffSchema(ctx, masterSchema, si.Shard.PrimaryAlias, alias, req.ExcludeTables, req.IncludeViews, &wg, &rec)
+	}
+	wg.Wait()
+	if rec.HasErrors() {
+		return nil, vterrors.Wrapf(rec.Error(), "schema diffs: %v", rec.Error().Error())
+	}
+	return &vtctldatapb.ValidateSchemaShardResponse{}, nil
+}
+
+// ValidateVSchema compares the schema of each primary tablet in "keyspace/shards..." to the vschema and errs if there are differences
+func (s *VtctldServer) ValidateVSchema(ctx context.Context, req *vtctldatapb.ValidateVSchemaRequest) (*vtctldatapb.ValidateVSchemaResponse, error) {
+	vschm, err := s.ts.GetVSchema(ctx, req.Keyspace)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetVSchema(%s) failed: %v", req.Keyspace, err)
+	}
+	var (
+		shardFailures concurrency.AllErrorRecorder
+		wg            sync.WaitGroup
+	)
+
+	wg.Add(len(req.Shards))
```

Comment:
you should add one at a time in the for loop to make the waitgroup usage easier to read